### ISSUE TITLE
Blocking the platform thread from a plugin causes UI to freeze

### ISSUE THREAD

**Issue by cbenhagen** [flutter/flutter#22024]
According to [the wiki](https://github.com/flutter/engine/wiki/Threading-in-the-Flutter-Engine#platform-task-runner) "...blocking the platform thread for inordinate amounts of time will not cause jank in Flutter applications,...".

Imagine a plugin with the following methods:
```objc
- (void)handleMethodCall:(FlutterMethodCall*)call result:(FlutterResult)result {
  if ([@"sleep" isEqualToString:call.method]) {
    [NSThread sleepForTimeInterval:10.0f];
    result(@"Good morning!");
  } else {
    result(FlutterMethodNotImplemented);
  }
}
```
```java
  @Override
  public void onMethodCall(MethodCall call, Result result) {
    if (call.method.equals("sleep")) {
      try {
        Thread.sleep(10000);
      } catch (InterruptedException e) {
        e.printStackTrace();
      }
      result.success("Good morning!");
    } else {
      result.notImplemented();
    }
  }
```

Calling these methods from Flutter causes the UI to freeze completely. As this happens both on Android and iOS this must be either by design or a bug. A full example project can be found here: https://github.com/cbenhagen/sleepy

```
[✓] Flutter (Channel dev, v0.9.1, on Mac OS X 10.14 18A389, locale de-CH)
    • Flutter version 0.9.1 at /Users/ben/flutter
    • Framework revision bf7c27095c (13 hours ago), 2018-09-18 13:39:21 -0700
    • Engine revision 9e315e6465
    • Dart version 2.1.0-dev.4.0.flutter-050561fd82

[✓] Android toolchain - develop for Android devices (Android SDK 27.0.3)
    • Android SDK at /Users/ben/Library/Android/sdk
    • Android NDK location not configured (optional; useful for native profiling support)
    • Platform android-27, build-tools 27.0.3
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_152-release-1024-b01)
    • All Android licenses accepted.

[✓] iOS toolchain - develop for iOS devices (Xcode 10.0)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 10.0, Build version 10A254a
    • ios-deploy 2.0.0
    • CocoaPods version 1.5.3

[✓] Android Studio (version 3.1)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin version 28.0.1
    • Dart plugin version 173.4700
    • Java version OpenJDK Runtime Environment (build 1.8.0_152-release-1024-b01)

[✓] Connected devices (2 available)
    • Android SDK built for x86 • emulator-5554                        • android-x86 • Android 9 (API 28) (emulator)
    • iPhone XS                 • 48A5A62A-7DB6-4C21-A7A9-961A089519D1 • ios         • iOS 12.0 (simulator)

• No issues found!
```

**Comment by cbenhagen** [comment]
@zoechi can you reevaluate the importance of this please? Either I am being very stupid here or this is a quite serious issue. Hoping for the former of course.

**Comment by YotaVGA** [comment]
I am experimenting exactly the same problem. I'd at least like to know if it is on purpose or not. If it is not the wiki should be a lot more clear than this, and I don't get why a future should be returned in this case...

**Comment by zoechi** [comment]
@cbenhagen the issue has not a single upvote so far https://github.com/flutter/flutter/wiki/Issue-hygiene#thumbs-up-reactions

**Comment by cbenhagen** [comment]
@zoechi I also wish people would use the upvote feature more but while having many upvotes is a good sign that an issue is important to many users, having no upvotes does not have to mean it can not be important. People writing plugins will just work around this issue by using threads or a dispatch_queue instead upvoting an issue. Like here for example: https://github.com/flutter/flutter/issues/25959

@YotaVGA please also give this issue your upvote.

**Comment by mehmetf** [comment]
My understanding is that this is by design:
- Plugins run on platform thread (basically UI thread of native OS)
- If you block UI thread, you will have a bad time (even though Flutter uses its own rendering thread). I am fairly certain gesture recognition would be affected since gestures flow through native UI thread first and has to be delegated to FlutterView. I am less sure whether rendering will be affected (for instance if you had a looping animation on Flutter). Perhaps @cbenhagen can test that for us.

@chinmaygarde as a fact checker emeritus here :-)

**Comment by chinmaygarde** [comment]
**TL;DR:** This is by design. The highlighted section from the wiki is part of a larger paragraph that provides more context and a specific recommendation against the example attached to this issue.

That sentence [on the wiki](https://github.com/flutter/flutter/wiki/The-Engine-architecture#platform-task-runner) is part of a larger discussion on platform threading behavior, the entirety of which I have pasted below:

> _"Even though (1)**blocking the platform thread for inordinate amounts of time will not cause jank in Flutter applications**(1), (2)**platforms do impose restrictions on expensive operations on this thread**(2). So (3)**it is advised that any expensive work in response to platform messages be performed on separate worker threads**(3) (unrelated to the four threads discussed above) before having the responses queued back on the the platform thread for submission to the engine. Not doing so **may result in platform-specific watchdogs terminating the application**(4)."_

Dissecting that paragraph,

1. From the perspective of the Flutter engine, the platform thread is like any other thread. None of the animation or frame workloads are processed on this thread. If you were to add a spinner to a Flutter application and block the platform thread, the spinner would keep spinning (without jank).
2. Platforms have their own restrictions on what you can do on the the main thread. Blocking this thread will prevent the system UI frameworks from delivering input events (causing your UI to freeze in response to touches) and in cases where this thread is blocked for a very long time, may even cause the system watchdog to terminate your application.
3. Blocking the platform thread via `Thread.sleep(10000);` counters the recommendation made on that wiki page. If you wanted to achieve this same effect, you could setup a timer on the platform thread and signal the same after the timeout. Or, more realistically, perform the expensive operation on a background thread.
4. As discussed, the platform is no longer sending pointer events (because it's main thread is blocked) and in extreme cases, you application may be terminated.

Suggestions on making the section in the wiki clearer are welcome of course.

**Comment by svaghar** [comment]
I have ran into exact same issue discussed here. The main design questions here is: why the platform plugin code is run on the main ui thread? If there is a convincing answer, then this is not a bug. Otherwise the current design forces developer to create a background thread.

**Comment by chinmaygarde** [comment]
The design was geared with the anticipation that plugins would be written to access platform specific APIs. The majority of such APIs on both iOS and Android are not thread safe and must be accessed from their "main" threads. The design didn't want to have authors of plugins immediately rethread all their platform plugin calls to the main thread.

For long running tasks, having the tasks run on a Flutter managed thread is also not very flexible. Would the engine have to maintain a thread or a pool or workers for parallel task execution? Besides , platforms already provide fairly ergonomic ways of dispatching tasks to worker threads (libDispatch on iOS for example).

Since the author of the plugin had the most knowledge about the kinds of tasks their plugin calls would make, this design was deemed most flexible and least surprising (to the majority of anticipated use cases).

**Comment by svaghar** [comment]
@cbenhagen thanks for explaining. couple of questions.

1) I read the The-Engine-architecture on flutter which explains 4 threads. I see two threads in my logcat: One running flutter/dart code and one thread running the java plugin code. Can you explain which one is which based on the engine arch?

2) My "long" running platfrom code are camera2 operations and filters. These could take 100s msecs to 1 second. What is the best way to design this? On a separate worker thread (e.g. Looper)?

**Comment by SergeShkurko** [comment]
@svaghar you can wrap you code with new thread. For example:
```java
 @Override
 public void onMethodCall(MethodCall call, Result result) {
  if (call.method.equals("sleep")) {
   new Thread(new Runnable() {
    public void run() {
     try {
      Thread.sleep(10000);
     } catch (InterruptedException e) {
      e.printStackTrace();
     }
     result.success("Good morning!");
    }
   }).start();
  } else {
   result.notImplemented();
  }
 }
```

**Comment by mrawl** [comment]
Nice try, but the callbacks (result) to dart are also supposed to be on the main thread AIUI. It would be nice if someone could clarify that.

**Comment by SergeShkurko** [comment]
@mrawl why do you think so? Everything works fine in multithreaded mode.

Here is a async callback examples (render function):
Android: https://github.com/rbcprolabs/flutter_plugins/blob/master/packages/native_pdf_renderer/android/src/main/java/io/scer/pdf/renderer/NativePDFRendererPlugin.kt
IOS: https://github.com/rbcprolabs/flutter_plugins/blob/master/packages/native_pdf_renderer/ios/Classes/SwiftNativePDFRendererPlugin.swift

**Comment by mrawl** [comment]
Yeh, my own code calls back on worker threads like your example, but check out this excellent document:
https://medium.com/flutter-io/flutter-platform-channels-ce7f540a104e

> Threads. Messages and replies are received, **and must be sent, on the platform’s main UI thread**. In Dart there is only one thread per Dart isolate, i.e. per Flutter view, so there is no confusion about which thread to use here.

Admittedly, it's not entirely clear whether he's referring to platform or dart side but I took it as both.

**Comment by p30arena** [comment]
I'm doing some bitmap and canvas,
but my flutter UI freezes,
whats happening?

I'm using the same pattern:

https://github.com/flutter/plugins/blob/c96eb1bf358a3d6e3cf7847bb921c114168210c1/packages/firebase_messaging/android/src/main/java/io/flutter/plugins/firebasemessaging/FirebaseMessagingPlugin.java#L133

**Comment by p30arena** [comment]
I want to draw your attention to this thread:

https://github.com/flutter/flutter/issues/30911

You should not call Result methods from a thread other than the main thread.###

**Comment by derolf** [comment]
I am initializing some stuff when the PlatformView is created at startup of the app, which might take some time. During that time, the app freezes.

Is that intended behavior?

**Comment by markfili** [comment]
I opted for coroutines in `onMethodCall`:
```
CoroutineScope(Dispatchers.Default).launch {
    val response = handleCall(call)
    launch(Dispatchers.Main) {
        result.success(response)
    }
}
```
added org.jetbrains.kotlinx:kotlinx-coroutines-core and kotlinx-coroutines-android to android folder's build.gradle

**Comment by Ehesp** [comment]
This can be achieved on Android using the [Task](https://developer.android.com/reference/com/google/android/play/core/tasks/Task) API fairly easily:

```java
 @Override
  public void onMethodCall(MethodCall call, @NonNull final MethodChannel.Result result) {

	Tasks.call(executor, () -> {
      // Do something expensive on a separate thread
      return "done";
    }).addOnCompleteListener(task -> {
     if (task.isSuccessful()) {
       result.success(task.getResult()); // "done"
     } else {
       result.error("error", task.getException())
     }
    });
```

The `executor` can then be some thread pool (see [Executors](https://developer.android.com/reference/java/util/concurrent/Executors)) - whatever you use depends on your performance & requirements.

**Comment by drriguz** [comment]
I have the same issue when implementing the [encryptions](https://pub.dev/packages/encryptions) plugin, I tried to use FlutterIsolate to wrap calls but it's super slower than before, so I ended using native threads.

In ios, a `DispatchQueue` is used:

```swift
let dispQueue = DispatchQueue(label: "com.riguz.encryptions");

// 
dispQueue.async {
someresult = ... // handle the heavy logic
DispatchQueue.main.async {
    result(someresult);
}
}
```

And in Android is pretty much the same idea, just use a new thread and `Looper.getMainLooper()`

I want to know if there is some official guide about this?

**Comment by darshankawar** [comment]
Closing this as working by design and per [this](https://github.com/flutter/flutter/issues/22024#issuecomment-465289341) and [this](https://github.com/flutter/flutter/issues/22024#issuecomment-470677255) comment. 

If you think this was closed by mistake and should be kept open, please write in comments and I'll reopen.
Thanks.
